homeassistant/components/mqtt/device_trigger.py

```diff
@@ -0,0 +1,285 @@
+    {
+        vol.Required(CONF_AUTOMATION_TYPE): str,
+        vol.Required(CONF_DEVICE): mqtt.MQTT_ENTITY_DEVICE_INFO_SCHEMA,
+        vol.Required(CONF_TOPIC): mqtt.valid_subscribe_topic,
+        vol.Optional(CONF_PAYLOAD, default=None): vol.Any(None, cv.string),
+        vol.Required(CONF_TYPE): cv.string,
+        vol.Required(CONF_SUBTYPE): cv.string,
+    },
+    mqtt.validate_device_has_at_least_one_identifier,
+)
+
+DEVICE_TRIGGERS = "mqtt_device_triggers"
+ATTACHED_DEVICE_TRIGGERS = "mqtt_attached_device_triggers"
+
+
+@attr.s(slots=True, frozen=True)
+class Trigger:
+    """Device trigger settings."""
+
+    device_id = attr.ib(type=str)
+    type = attr.ib(type=str)
+    subtype = attr.ib(type=str)
+    topic = attr.ib(type=str)
+    payload = attr.ib(type=str)
+    qos = attr.ib(type=int)
+    discovery_id = attr.ib(type=str)
+
+
+@attr.s(slots=True)
+class AttachedTrigger:
+    """Device trigger settings."""
+
+    device_id = attr.ib(type=str)
+    discovery_id = attr.ib(type=str)
+    action = attr.ib(type=AutomationActionType)
+    automation_info = attr.ib(type=dict)
+    remove = attr.ib(type=CALLBACK_TYPE)
+
+
+async def _update_device(hass, config_entry, config):
+    """Update device registry."""
+    device_registry = await hass.helpers.device_registry.async_get_registry()
+    config_entry_id = config_entry.entry_id
+    device_info = mqtt.device_info_from_config(config[CONF_DEVICE])
+
+    if config_entry_id is not None and device_info is not None:
+        device_info["config_entry_id"] = config_entry_id
+        device_registry.async_get_or_create(**device_info)
+
+
+async def _attach_mqtt_trigger(hass, discovery_id, action, automation_info):
+    """Attach MQTT trigger."""
+    trigger = hass.data[DEVICE_TRIGGERS][discovery_id]
+    mqtt_config = {
+        automation_mqtt.CONF_TOPIC: trigger.topic,
+        automation_mqtt.CONF_ENCODING: DEFAULT_ENCODING,
+        automation_mqtt.CONF_QOS: trigger.qos,
+    }
+    if trigger.payload:
+        mqtt_config[CONF_PAYLOAD] = trigger.payload
+
+    return await automation_mqtt.async_attach_trigger(
+        hass, mqtt_config, action, automation_info
+    )
+
+
+async def _update_trigger(hass, discovery_id, config):
+    """Remove MQTT device trigger."""
+    device_id = hass.data[DEVICE_TRIGGERS][discovery_id].device_id
+    hass.data[DEVICE_TRIGGERS][discovery_id] = Trigger(
+        device_id=device_id,
+        type=config[CONF_TYPE],
+        subtype=config[CONF_SUBTYPE],
+        topic=config[CONF_TOPIC],
+        payload=config[CONF_PAYLOAD],
+        qos=config[CONF_QOS],
+        discovery_id=discovery_id,
+    )
+
+    if ATTACHED_DEVICE_TRIGGERS not in hass.data:
+        return
+
+    # Unsubscribe+subscribe if this trigger is in use
+    for trig in hass.data[ATTACHED_DEVICE_TRIGGERS]:
+        if trig.discovery_id == discovery_id:
+            if trig.remove:
+                trig.remove()
+            trig.remove = await _attach_mqtt_trigger(
```

Comment:
This feels like logic that should live inside the Trigger class. Right now you've made the trigger class just some properties, but it could easily have logic like "attach" and "update config" (which detaches and attaches new one if it is currently attached)

Reply:
```python
class Trigger:
    def __init__(self, hass)
    async def attach(action)
    async def detach()
    async def update_config()
```

Reply:
Good point, refactored accordingly.